### PR TITLE
Block Host with Cisco ASA from Microsoft Teams

### PR DESCRIPTION
**Pull request by r7-kszczepanskagorna**
## Requirements

Workflow contributors, verify you have completed the following items by checking them off:

### Checklist
- [ ] Workflow verified as working in a test
- [ ] Workflow bundle includes `Name_of_Workflow.icon` file, `help.md` file, `workflow.spec.yaml` file, and `screenshots` directory
- [ ] Workflow title should follow general [Title of Articles](http://grammar.yourdictionary.com/capitalization/rules-for-capitalization-in-titles.html) format
- [ ] Workflow includes a detailed Description
- [ ] Workflow step names follow [Title of Articles](http://grammar.yourdictionary.com/capitalization/rules-for-capitalization-in-titles.html) format
- [ ] Triggers, Decisions, and other workflow steps include descriptions where needed
- [ ] Workflow includes at least one `Final Report` artifact that summarizes workflow execution details
- [ ] Workflow bundle (directory) is named after the title of the workflow in snake case with first letters capitalized, e.g. `workflows/Name_of_Workflow/`
- [ ] `help.md` file includes Description, Key Features, Requirements, Setup, Technical Details, Version History, and References
- [ ] `workflow.spec.yaml` file includes `products`, `name`, `title`, `description`, `use_cases` tags, `keywords` tags, `source_url`, and the `name` for each screenshot matches a PNG image in the `screenshots` directory

### Assessment

```
[*] Validating workflow at .

[*] Running Integration Validators...
[*] Executing validator WorkflowDirectoryNameMatchValidator
[*] Executing validator WorkflowFilesValidator
[*] Executing validator WorkflowHelpValidator
[*] Executing validator WorkflowChangelogValidator
[*] Executing validator WorkflowVendorValidator
[*] Executing validator WorkflowVersionValidator
[*] Executing validator WorkflowExtensionValidator
[*] Executing validator WorkflowSupportValidator
[*] Executing validator WorkflowPNGHashValidator
[*] Executing validator WorkflowICONFileNameValidator
[*] Executing validator WorkflowScreenshotValidator
[*] Executing validator WorkflowTitleValidator
[*] Executing validator WorkflowDescriptionValidator
[*] Executing validator WorkflowNameValidator
[*] Executing validator WorkflowProfanityValidator
[*] Executing validator WorkflowHelpPluginUtilizationValidator
[*] Executing validator WorkflowICONFileValidator
[*] Workflow successfully validated!

----
[*] Total time elapsed: 87.062ms
```

<summary>
icon-validate 
</summary>

#### Workflow Builder

workflow1

workflow2

workflow3

#### Job

teams1

teams2

teams3

#### Artifact

```

### CODE REVIEW

**Review by jmcadams-r7 — APPROVED** [review]

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+# Description
+
+This workflow blocks or unblocks a IPv4 and IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:3]:
```suggestion
This workflow blocks or unblocks IPv4 and IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+## Technical Details
+
+Plugins utilized by workflow:
+
+|Plugin|Version|Count|
+|----|----|--------|
+|Cisco Adaptive Security Appliance|1.4.1|5|
+|Type Converter|1.6.0|1|
+|Microsoft Teams|2.2.1|8|
+
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:53]: THere's an extra newline here

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+After configuring the Microsoft Teams steps, activate the workflow in order to trigger it.
+
+
+### Usage
+
+*This workflow will only trigger in the channel specified in the Microsoft Teams workflow steps.*
+
+To run the workflow, send a message to the specified Microsoft Teams channel starting with the command `!block-host` or `!unblock-host`.
+
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:40]: THere's an extra newline here

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+|Plugin|Version|Count|
+|----|----|--------|
+|Cisco Adaptive Security Appliance|1.4.1|5|
+|Type Converter|1.6.0|1|
+|Microsoft Teams|2.2.1|8|
+
+
+## Troubleshooting
+
+_There is no troubleshooting information at this time_
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:57]:
```suggestion
The REST API server must be installed and enabled on the Cisco ASA device this plugin connects to. Cisco provides online documentation for this available [here](https://www.cisco.com/c/en/us/td/docs/security/asa/api/qsg-asa-api.html). In addition, the user account must have the necessary permissions for the intended actions:

* Actions that retrieve or check data require Cisco privilege level 5 or greater
* Actions that change or modify data require Cisco privilege level 15
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/workflow.spec.yaml [hunk 5]
```diff
@@ -0,0 +1,40 @@
+extension: workflow
+products:
+- insightconnect
+name: Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams
+title: Block Host with Cisco ASA Firewall from Microsoft Teams
+description: This workflow blocks or unblocks a IPv4 and IPv6 addresses with Cisco ASA via Microsoft Teams
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/workflow.spec.yaml:6]:
```suggestion
description: This workflow blocks or unblocks IPv4 and IPv6 addresses with Cisco ASA via Microsoft Teams
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 6]
```diff
@@ -0,0 +1,3143 @@
...
+              "imageData": "https://us.cdn-assets.connect.insight.rapid7.com/icons/rapid7/microsoft_teams/2.2.1/icon.png"
+            },
+            "identifier": "send_message",
+            "continueOnFailure": false,
+            "isDisabled": false,
+            "isCloud": false,
+            "parameters": {
+              "input": {
+                "channel_name": "{{[ffa084e7-7b4d-4972-8ad8-f13f2877b669].[channel_name]}}",
+                "message": "⚠️ Cisco ASA support only IPv4 and IPv6 addresses. Check address input and try again. ⚠️",
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:2221]:
```suggestion
                "message": "⚠️ Cisco ASA only suppors IPv4 and IPv6 addresses. Check the address input and try again. ⚠️",
```

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]
Capitalization isn't consistent in step names - there's a quite a few.

e.g. `IP address unblocked` should be `IP Address Unblocked`
        `Unblock IP addresses` should be `Unblock IP Addresses`

**Comment by jschipp-r7** [comment]
Let's remove the API screenshots too.

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 7]
```diff
@@ -0,0 +1,3143 @@
...
+  "kom": {
+    "komandVersion": "1.65.0-0-g6376e",
+    "komFileVersion": "2.0.0",
+    "exportedAt": "2020-10-19T16:12:39.490521778Z",
+    "workflowVersions": [
+      {
+        "name": "Block Host with Cisco ASA Firewall from Microsoft Teams",
+        "type": "runnable",
+        "version": "",
+        "description": "This workflow blocks or unblocks an IPv4 or IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams .",
```
Comment (mrinehart-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:11]:
```suggestion
        "description": "This workflow blocks or unblocks an IPv4 or IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.",
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+# Documentation
+
+## Setup
+
+Import the workflow from the Rapid7 Extension Library and proceed through the Import Workflow wizard in InsightConnect. Import plugins, create or select connections, and rename the workflow as a part of the Import Workflow wizard as necessary.
+
+Once the workflow has been imported, **Update the first step with the channel name to suit your Microsoft Teams environment!** by editing the input with the preset text of `change_me` to match the channel to monitor.
+
+After configuring the Microsoft Teams steps, activate the workflow in order to trigger it.
+
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:33]:
```suggestion
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 9]
```diff
@@ -0,0 +1,3143 @@
...
+              "imageData": "https://us.cdn-assets.connect.insight.rapid7.com/icons/rapid7/cisco_asa/1.4.1/icon.png"
+            },
+            "identifier": "add_address_to_group",
+            "continueOnFailure": true,
+            "isDisabled": false,
+            "isCloud": false,
+            "parameters": {
+              "input": {
+                "address": "{{[3e3020b6-0ae4-4739-b40b-9ee2a854e716].[$item]}}",
+                "group": "InsightConnect_Block_List"
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:988]:
```suggestion
                "group": "change_me"
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 10]
```diff
@@ -0,0 +1,3143 @@
...
+            },
+            "identifier": "check_if_address_object_in_group",
+            "continueOnFailure": true,
+            "isDisabled": false,
+            "isCloud": false,
+            "parameters": {
+              "input": {
+                "address": "{{[3e3020b6-0ae4-4739-b40b-9ee2a854e716].[$item]}}",
+                "enable_search": false,
+                "group": "InsightConnect_Block_List"
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:1411]:
```suggestion
                "group": "change_me"
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 11]
```diff
@@ -0,0 +1,3143 @@
...
+              "imageData": "https://us.cdn-assets.connect.insight.rapid7.com/icons/rapid7/cisco_asa/1.4.1/icon.png"
+            },
+            "identifier": "remove_address_from_group",
+            "continueOnFailure": true,
+            "isDisabled": false,
+            "isCloud": false,
+            "parameters": {
+              "input": {
+                "address": "{{[a18318ff-f4ce-432b-8274-2f46b977d066].[$item]}}",
+                "group": "InsightConnect_Block_List"
```
Comment (jschipp-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:2450]:
```suggestion
                "group": "change_me"
```

**Review by jschipp-r7 — CHANGES_REQUESTED** [review]
@r7-kszczepanskagorna can you make the following changes from the Slack PR here
https://github.com/rapid7/insightconnect-workflows/pull/355

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon [hunk 12]
```diff
@@ -0,0 +1,3143 @@
...
+  "kom": {
+    "komandVersion": "1.65.0-0-g6376e",
+    "komFileVersion": "2.0.0",
+    "exportedAt": "2020-10-19T16:12:39.490521778Z",
+    "workflowVersions": [
+      {
+        "name": "Block Host with Cisco ASA Firewall from Microsoft Teams",
+        "type": "runnable",
+        "version": "",
+        "description": "This workflow blocks or unblocks an IPv4 or IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.",
```
Comment (mrinehart-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams.icon:11]:
```suggestion
        "description": "This workflow blocks or unblocks IPv4 or IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.",
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
+# Description
+
+This workflow blocks or unblocks IPv4 and IPv6 addresses with Cisco ASA via Microsoft Teams command and reports information back to Microsoft Teams.
+The workflow works by adding and removing address objects from an existing network group tied to a Deny firewall policy in your Cisco ASA Firewall.
```
Comment (mrinehart-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:4]:
```suggestion

The workflow works by adding and removing address objects from an existing network group tied to a Deny firewall policy in your Cisco ASA Firewall.
```

File: workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+When an address is added to the group, it's blocked; when an address is removed from a group, it's unblocked. These are best practices for automating the blocking of hosts.
```
Comment (mrinehart-r7) [inline workflows/Block_Host_with_Cisco_ASA_Firewall_from_Microsoft_Teams/help.md:5]:
```suggestion

When an address is added to the group, it's blocked; when an address is removed from a group, it's unblocked. These are best practices for automating the blocking of hosts.
```

**Review by mrinehart-r7 — CHANGES_REQUESTED** [review]

**Review by mrinehart-r7 — APPROVED** [review]

**Pull request merged**